File: Projects/Delphi/QixAlgorithm/UMap.pas
unit UMap;

interface

{* This program was downloaded from http://classic-games.pl.
 * It's an example of filling algorithm as seen in the game Qix or Volfied.
 *
 * You may use this code in any way but if you create any application or game using
 * this code (a part or the whole) you should mention about the author of original code,
 * that is Grzegorz Drozd - and the source website http://classic-games.pl.
 * That's all. If you want, that will be an honour for me if you write to me and tell
 * about it! Have fun.
 *
 * (c) 2019-04-14 Grzegorz Drozd,
 * http://classic-games.pl, https://gadz.pl, gad@gadz.pl
 *}

uses
  System.SysUtils,
  System.Classes,
  System.Generics.Collections,
  WinApi.Windows;

type
  TMapPointType = (pEmpty, pLine, pFilled);

  TMapPoint = class
  public
    Value: TMapPointType;
    Occupied: boolean;
    id: integer;

    constructor Create;
  end;

  TMapRow = TObjectList<TMapPoint>;
  TMapMatrix = TObjectList<TMapRow>;

  TMap = class
  private
    Width, Height: integer;
    points: TMapMatrix;

    procedure CheckPosition(px, py: integer);
  public
    constructor Create;
    destructor Destroy; override;
    procedure Make(sx, sy: integer);
    function GetPoint(px, py: integer): TMapPointType;
    procedure SetPoint(px, py: integer; Value: TMapPointType);
    function GetOccupied(px, py: integer): boolean;
    procedure SetOccupied(px, py: integer; Value: boolean);
    function GetId(px, py: integer): integer;
    procedure SetId(px, py: integer; Value: integer);
    procedure DrawHLine(x1, y, x2: integer; Value: TMapPointType);
    procedure DrawVLine(x, y1, y2: integer; Value: TMapPointType);

    function GetWidth: integer;
    function GetHeight: integer;
  end;

implementation

procedure Swap(var v1, v2: integer);
var
  tmp: integer;
begin
  tmp := v1;
  v1 := v2;
  v2 := tmp;
end;

{ TMapPoint }

constructor TMapPoint.Create;
begin
  inherited;
  Value := TMapPointType.pEmpty;
  Occupied := false;
  id := 0;
end;

{ TMap }

procedure TMap.CheckPosition(px, py: integer);
begin
  if (px < 0) or (py < 0) or (px >= Self.Width) or (py >= Self.Height)  then
    raise Exception.CreateFmt('Map point out of range [%dx%d]', [px, py]);
end;

constructor TMap.Create;
begin
  inherited;
  points := TMapMatrix.Create(true);
end;

destructor TMap.Destroy;
begin
  points.Free;
  inherited;
end;

procedure TMap.DrawHLine(x1, y, x2: integer; Value: TMapPointType);
var
  x: integer;
begin
  if x1 > x2 then
    Swap(x1, x2);
  for x := x1 to x2 do
    SetPoint(x, y, Value);
end;

procedure TMap.DrawVLine(x, y1, y2: integer; Value: TMapPointType);
var
  y: integer;
begin
  if y1 > y2 then
    Swap(y1, y2);
  for y := y1 to y2 do
    SetPoint(x, y, Value);
end;

procedure TMap.Make(sx, sy: integer);
var
  x, y: integer;
  MapRow: TMapRow;
begin
  self.Width := sx;
  self.Height := sy;

  points.Clear;
  for y := 0 to sy - 1 do
  begin
    MapRow := TMapRow.Create(true);
    for x := 0 to sx - 1 do
      MapRow.Add(TMapPoint.Create);
    points.Add(MapRow);
  end;

end;

procedure TMap.SetId(px, py, Value: integer);
begin
  CheckPosition(px, py);
  points[py][px].id := Value;
end;

procedure TMap.SetOccupied(px, py: integer; Value: boolean);
begin
  CheckPosition(px, py);
  points[py][px].Occupied := Value;
end;

procedure TMap.SetPoint(px, py: integer; Value: TMapPointType);
begin
  CheckPosition(px, py);
  points[py][px].Value := Value;
end;

function TMap.GetHeight: integer;
begin
  result := Height;
end;

function TMap.GetId(px, py: integer): integer;
begin
  CheckPosition(px, py);
  result := points[py][px].id;
end;

function TMap.GetOccupied(px, py: integer): boolean;
begin
  CheckPosition(px, py);
  result := points[py][px].Occupied;
end;

function TMap.GetPoint(px, py: integer): TMapPointType;
begin
  CheckPosition(px, py);
  result := points[py][px].Value;
end;

function TMap.GetWidth: integer;
begin
  result := Height;
end;

end.

File: Projects/C#/QixAlgorithm/UMap.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

/* This program was downloaded from http://classic-games.pl.
 * It's an example of filling algorithm as seen in the game Qix or Volfied.
 * 
 * You may use this code in any way but if you create any application or game using
 * this code (a part or the whole) you should mention about the author of original code,
 * that is Grzegorz Drozd - and the source website http://classic-games.pl.
 * That's all. If you want, that will be an honour for me if you write to me and tell
 * about it! Have fun.
 * 
 * (c) 2019-04-14 Grzegorz Drozd, 
 * http://classic-games.pl, https://gadz.pl, gad@gadz.pl
 */


namespace QixAlgorithm
{
    public enum TMapPointType { pEmpty, pLine, pFilled };

    public class TMapPoint {

        public TMapPointType Value;
        public bool Occupied;
        public int id;

        public TMapPoint()
        {
            Value = TMapPointType.pEmpty;
            Occupied = false;
            id = 0;
        }
    }

    public class TMapRow : List<TMapPoint> { };
    public class TMapMatrix : List<TMapRow> { };

    public class TMap {
        private int Width, Height;
        private TMapMatrix points;

        private void Swap(ref int v1, ref int v2)
        {
            int tmp;
            tmp = v1;
            v1 = v2;
            v2 = tmp;
        }

        private void CheckPosition(int px, int py) {
            if (px < 0 || py < 0 || px >= this.Width || py >= this.Height)
                throw new Exception(String.Format("Map point out of range [{0}x{1}]", px, py));
        }

        public TMap()
        {
            points = new TMapMatrix();
        }

        public void DrawHLine(int x1, int y, int x2, TMapPointType Value)
        {
            if (x1 > x2)
                Swap(ref x1, ref x2);
            for (int x = x1; x <= x2; x++)
                SetPoint(x, y, Value);
        }

        public void DrawVLine(int x, int y1, int y2, TMapPointType Value)
        {
            if (y1 > y2)
                Swap(ref y1, ref y2);
            for (int y = y1; y <= y2; y++)
                SetPoint(x, y, Value);
        }

        public void Make(int sx, int sy)
        {

            TMapRow MapRow;
            this.Width = sx;
            this.Height = sy;

            points.Clear();
            for (int y = 0; y <= sy - 1; y++)
            {
                MapRow = new TMapRow();
                for (int x = 0; x <= sx - 1; x++)
                    MapRow.Add(new TMapPoint());

                points.Add(MapRow);
            }

        }

        public void SetId(int px, int py, int Value)
        {
            CheckPosition(px, py);
            points[py][px].id = Value;
        }

        public void SetOccupied(int px, int py, bool Value)
        {
            CheckPosition(px, py);
            points[py][px].Occupied = Value;
        }

        public void SetPoint(int px, int py, TMapPointType Value)
        {
            CheckPosition(px, py);
            points[py][px].Value = Value;
        }

        public int GetHeight()
        {
            return Height;
        }

        public int GetId(int px, int py)
        {
            CheckPosition(px, py);
            return points[py][px].id;
        }

        public bool GetOccupied(int px, int py)
        {
            CheckPosition(px, py);
            return points[py][px].Occupied;
        }

        public TMapPointType GetPoint(int px, int py)
        {
            CheckPosition(px, py);
            return points[py][px].Value;
        }

        public int GetWidth()
        {
            return Width;
        }

    }

}
